### ISSUE TITLE
Jank on every query

### ISSUE THREAD

**Issue by budde377** [zino-hofmann/graphql-flutter#1196]
### Discussed in https://github.com/zino-hofmann/graphql-flutter/discussions/1195

<div type='discussions-op-text'>

<sup>Originally posted by **aquadesk** July 23, 2022</sup>
Hi all,

I have a list page and tapping one of them opens a detail view page

On every detail view load, it comes with a jank

I am not certain about the root cause of the janks

But what I see on the dev tools, it seems it's coming from managing cache.

In my stateful hook widget, i do

```

    final reviewHook = useQuery$ProductReview(Options$Query$ProductReview(
      variables: Variables$Query$ProductReview(
        id: widget.productReview.id!,
      ),      
      fetchPolicy: FetchPolicy.cacheAndNetwork,
    ));

```

![image](https://user-images.githubusercontent.com/19755104/180581638-2b62bb0d-ccea-4d00-882f-2a7b91a1a327.png)

![image](https://user-images.githubusercontent.com/19755104/180581862-cd845e9a-eb12-4a08-971e-3e279b4c6445.png)

It seems it takes 2.5s to run a query and process it on a device (one plus 6) in a profile mode.

I am not a professional flutter engineer so please excuse me for wrong terms or assumptions.

I can provide more context or code snippet how I do things on my app but not sure how much I should share to give a proper amount of context for this. 

I've tried almost everything that could contribute to this jank but I see that as soon as the app makes a graphql query, it janks. 
Even with the Fetch policy noCache.

I was using 5.1.1. beta 2

Thanks for the awesome project!

Any idea, direction or suggestion how I should go about figuring out what is causing the jank would be appreciated.</div>

**Comment by orestesgaolin** [comment]
Typically the json deserialization is a synchronous process and in case of big json objects it can take significant amount of time. What's more or less the size of the response you receive?

**Comment by budde377** [comment]
I don't think this is caused by serialization, but rather how we compare if the data has changed, but it amounts to the same issue. 

@aquadesk - until we can revisit isolates here, it might make sense to skip this comparison and just always rebroadcast? This is expensive if you have a lot of queries mounted in your app and a lot of updates, but might just work in your case. 

You can disable the comparison by setting `alwaysRebroadcast: true` on your client.

**Comment by orestesgaolin** [comment]
<img width="978" alt="CleanShot 2022-08-15 at 16 15 10@2x" src="https://user-images.githubusercontent.com/16854239/184662777-605308dc-236d-4b18-a400-a2b4560c15d2.png">

When profiling with the current stable version I noticed significantly long time spent on the `decode` from dart:convert. With the new beta version `0.11.0-beta.5` this time seems significantly shorter.

**Comment by orestesgaolin** [comment]
Is there any chance to move the deserialization to another isolate? Some libraries like objectbox were able to offload this process and with more and more data coming in to the app, I'd love to be able to benefit from the same approach.

**Comment by budde377** [comment]
I think this is an interesting idea. IIRC the deserialisation is done in the link which uses a stream based api, so you should be able to quite easily offload this into another isolate.

Do you have capacity to give this a go?

**Comment by orestesgaolin** [comment]
Could you point me to a class or file that would be a good start? Happy to look into that and maybe convince the client to sponsor it ;)

**Comment by budde377** [comment]
If you take a look at the [client](packages/graphql/lib/src/graphql_client.dart) you can provide a custom implementation of the link instead of using a standard http link.

The link could act as a proxy to the http link, so all you need is to initialise and send the request/response to that isolate and eventually the underlying http link.

**Comment by orestesgaolin** [comment]
I was looking into the client implementation and have couple of questions. I noticed it's possible for `HttpLink` to override `httpResponseDecoder` like this:

```dart
Future<Map<String, dynamic>?> isolateHttpResponseDecoder(
  http.Response httpResponse,
) async =>
    await compute(jsonDecode, httpResponse.body) as Map<String, dynamic>?;

final baseLink = HttpLink(
          baseUrl,
          httpResponseDecoder: isolateHttpResponseDecoder,
        );
```

However, for websocket connections I cannot find similar pattern. There's a possibility to provide custom `ResponseParser` to `SocketClientConfig`, but it seems too late to benefit from separate isolate, as `Response parseResponse(Map<String, dynamic> body)` accepts the Map which must have been already decoded at this stage.

What I've found is that `GraphQLSocketMessage` in the method `parse` uses `json.decode`

https://github.com/zino-hofmann/graphql-flutter/blob/312dc6cb60679b588db3889054219ef249f7c561/packages/graphql/lib/src/links/websocket_link/websocket_messages.dart#L56-L62

It seems that we would benefit from using `compute` in this method, but I'm not sure if you agree. Would there be a way to override this behavior?

**Comment by orestesgaolin** [comment]
Small update from my side. I was able to use `compute` function from [this package](https://github.com/dartsidedev/compute) like that:

```dart
import 'package:compute/compute.dart';

...

class GraphQLWebSocketChannel extends StreamChannelMixin<dynamic>
    implements WebSocketChannel {
...

  /// Stream of messages from the endpoint parsed as GraphQLSocketMessages
  Stream<GraphQLSocketMessage> get messages =>
      _messages ??= stream.asyncMap<GraphQLSocketMessage>(
        (d) async => await compute(GraphQLSocketMessage.parse, d),
      );

...
}
```

Here are isolates that get spawned when getting new messages via the websocket


https://user-images.githubusercontent.com/16854239/193909575-2848f0bf-965e-4948-b060-4ea7dcc9244c.mp4


The cool thing is that now I no longer notice jank while scrolling in the app

**Comment by orestesgaolin** [comment]
For anyone interested here's my WS link that operates on isolates when decoding json messages. If you spot any mistakes, please let me know.

```dart
// ignore_for_file: depend_on_referenced_packages

import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:graphql/client.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

class CustomWsLink extends WebSocketLink {
  CustomWsLink(
    super.url, {
    super.config,
    super.subProtocol,
    required this.getToken,
    required this.userUid,
  });

  final Future<String> Function() getToken;
  final String userUid;
  _Connection? _connection;

  /// this will be called every time you make a subscription
  @override
  // ignore: type_annotate_public_apis
  Stream<Response> request(Request request, [forward]) async* {
    final token = await getToken();

    /// check is connection is null or the token changed
    if (_connection == null || _connection!.token != token) {
      _connectOrReconnect(token);
    }
    yield* _connection!.client.subscribe(request, true);
  }

  /// Connects or reconnects to the server with the specified headers.
  void _connectOrReconnect(String token) {
    _connection?.client.dispose();

    _connection = _Connection(
      client: SocketClient(
        url,
        config: SocketClientConfig(
          // ignore: avoid_redundant_argument_values
          autoReconnect: true,
          connectFn: (uri, protocols) {
            var channel = WebSocketChannel.connect(uri, protocols: protocols);
            // ignore: join_return_with_assignment
            channel = channel.forGraphQLIsolate();

            return channel;
          },
          initialPayload: () => {
            'headers': {
              'Authorization': token,
              'Content-Type': 'application/json',
            },
          },
        ),
      ),
      token: token,
    );
  }

  @override
  Future<void> dispose() async {
    await _connection?.client.dispose();
    _connection = null;
  }
}

/// this a wrapper for web socket to hold the used token
class _Connection {
  _Connection({
    required this.client,
    required this.token,
  });

  final SocketClient client;
  final String token;
}

class CustomGraphQLWebSocketChannel extends GraphQLWebSocketChannel {
  CustomGraphQLWebSocketChannel(super.webSocket);

  Stream<GraphQLSocketMessage>? _messages;

  /// Stream of messages from the endpoint parsed as GraphQLSocketMessages
  @override
  Stream<GraphQLSocketMessage> get messages =>
      _messages ??= stream.asyncMap<GraphQLSocketMessage>(
        (d) async => compute(GraphQLSocketMessage.parse, d),
      );
}

extension GraphQLGetter on WebSocketChannel {
  /// Returns a wrapper that has safety and convenience features for graphql
  GraphQLWebSocketChannel forGraphQLIsolate() =>
      this is CustomGraphQLWebSocketChannel
          ? this as CustomGraphQLWebSocketChannel
          : CustomGraphQLWebSocketChannel(this);
}
```

**Comment by mkustermann** [comment]
The current implementation of `_cachedDataHasChangedFor` is based on `package:collection`s `DeepCollectionEquality().equals` implementation. It seems this implementation is extremely slow. See https://github.com/dart-lang/collection/issues/263. You may consider using a different implementation

**Comment by vincenzopalazzo** [comment]
Thank @mkustermann It looks like it is the case to digger into it with more passion now that we know that the `DeepCollectionEquality().equals` call is unoptimal by design

**Comment by insidewhy** [comment]
Surprised the priority for this is low given any app that uses this library is janky even on high powered phones, and the level of jank raises to intolerable for mid-to-low powered phones.

At my company we're attempting to port away from this library and to https://ferrygraphql.com/ which does all significant work in an isolate, I'll report back here on how well this works.

**Comment by jeff-come-on** [comment]
> _cachedDataHasChangedFor

We're noticing huge lags in our messaging app when using writeFragment, any performance boost using ferry ?

**Comment by JWambaugh** [comment]
I am experiencing severe jank seemingly due to this in my app. I can't understand why this is a low priority since it is affecting performance so badly, especially on the web or slower phones. Looks like I might have to port my app to ferry as well?

**Comment by insidewhy** [comment]
> > _cachedDataHasChangedFor
> 
> We're noticing huge lags in our messaging app when using writeFragment, any performance boost using ferry ?

After porting to ferry we were very happy, our users stopped complaining that our app felt slow.
